infra/modules/azuread-connection/main.tf

```diff
@@ -0,0 +1,39 @@
+# provisions infra for a Microsoft Data Source connector (in Azure AD)
+#  - the connector application
+#  - granting connector access on behalf of the users in your Azure AD directory
+
+terraform {
+  required_providers {
+    azuread = {
+      version = "~> 2.15.0"
+    }
+  }
+}
+
+data "azuread_application_published_app_ids" "well_known" {}
```

Comment:
Main problem of this, as Terraform docs says as that it comes from an *unofficial* list. 

The id of `MicrosoftGraph` application has been always the same as far I remember (`00000003-0000-0000-c000-000000000000`). Of course it could change without warning in the future, but note that same app is not only used for any third party you could imagine but also for a lot of MSFT services.

There is a [link with official Ids](https://docs.microsoft.com/en-us/troubleshoot/azure/active-directory/verify-first-party-apps-sign-in#application-ids-for-commonly-used-microsoft-applications) that may be more accurate to refer on this case.

For this, I'd consider to have `application_id` in L16 as constant with a comment on the official documentation.

Reply:
yeah .... not sure what to do here. That the terraform code is doing is, in effect, the same as what MSFT tells you to do to look up permission IDs (https://docs.microsoft.com/en-us/graph/permissions-reference#retrieving-permission-ids) - except that they're querying by `appDisplayName =='Microsoft Graph'` . imho, that seems like a worse practice, as that's nothing but a marketing name

Reply:
Just FYI yeap, that's not a good practice. Display name can be repeated and the safest way is to use the proper GUID because is the only unique value (id) that identifies an AAD application. 

I've just created a couple AAD applications with an SP named `Microsoft Graph` in my personal AAD:

![image](https://user-images.githubusercontent.com/75276364/151146543-55d8b2a4-a45a-4daa-a980-35b88590ad73.png)

If we know the GUID best approach is to use it. ID will not change unless the application is destroyed, something that in this case is hard to happen. But if we don't know the ID and we need to perform a lookup as the MSFT docs says, we need more else to check -as they say, checking things in the properties as permissions, paths, whatever- and then order them by their expiration date of their keys or something, because we *could* have more than one entry and then get the wrong ID to use.